tournamentcontrol/competition/tests/test_utils.py

```diff
@@ -264,25 +265,201 @@ def test_edit_stage_view_with_undecided_teams_referencing_missing_pools(self):
+        # Note: Don't set both label and formula as per UndecidedTeamForm validation
+        undecided_team_g5p2 = UndecidedTeamFactory.create(
+            stage=self.finals_stage,
+            formula="G5P2",  # References group 5, but we only have 2 pools
+        )
+
+        undecided_team_s1g5p1 = UndecidedTeamFactory.create(
+            stage=self.finals_stage,
+            formula="S1G5P1",  # Explicit stage reference, but group 5 doesn't exist
+        )
+
+        # Test the core fix: accessing title should not raise IndexError
+        # After the fix, invalid formulas should return the formula as fallback
+        self.assertEqual(undecided_team_g5p2.title, "G5P2")
+        self.assertEqual(undecided_team_s1g5p1.title, "S1G5P1")
+
+        # Test that choices also work correctly (fallback to division teams)
+        self.assertEqual(undecided_team_g5p2.choices, self.finals_stage.division.teams)
+        self.assertEqual(undecided_team_s1g5p1.choices, self.finals_stage.division.teams)
+
+    def test_invalid_formulas_in_match_eval_fields(self):
+        """
+        Test that invalid formulas in home_team_eval and away_team_eval fields
+        are handled gracefully when displayed in admin views and when Match.eval() is called.
+        """
+        # Create a Match with invalid formulas in eval fields
+        match_with_invalid_formulas = MatchFactory.create(
+            stage=self.finals_stage,
+            home_team=None,
+            away_team=None,
+            home_team_eval="G5P2",  # Invalid - references group 5, but only 2 pools exist
+            away_team_eval="G3P1",  # Invalid - references group 3, but only 2 pools exist
+            label="Invalid Formula Match",
+        )
+
+        # Test that accessing the match in admin view doesn't crash
+        self.login(self.user)
+        self.get(self.finals_stage.urls["edit"])
+        self.response_200()
+        # The match should be displayed on the stage edit page
+        self.assertResponseContains("Invalid Formula Match")
+
+        # Test the eval() method directly - it should handle the IndexError gracefully
+        # The Match.eval() method should not raise IndexError even with invalid formulas
+        # Teams will be None if formula can't be resolved, which is expected behavior
+        # The key is that no IndexError should be raised
+        self.assertIsNotNone(match_with_invalid_formulas.eval())
+
+        # Also test that get_home_team() and get_away_team() don't crash
+        # These methods call _get_team() which should handle invalid formulas gracefully
+        # The results should be dictionaries with title indicating the error
+        # or None if the evaluation couldn't be completed
+        self.assertIsNotNone(match_with_invalid_formulas.get_home_team())  # Should return something, not crash
+        self.assertIsNotNone(match_with_invalid_formulas.get_away_team())  # Should return something, not crash
```

Comment:
As advised earlier, don't assert that something is "anything but `None`" because it could still be incorrectly evaluated.

Always assert positively - in this case, `assertEqual` that the value returned from the method call is what we expect.